pkg/bgp/speaker/speaker_test.go

```diff
@@ -0,0 +1,482 @@
+package speaker
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/cilium/cilium/pkg/bgp/mock"
+	"github.com/cilium/cilium/pkg/k8s"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/google/go-cmp/cmp"
+
+	metallbbgp "go.universe.tf/metallb/pkg/bgp"
+	"go.universe.tf/metallb/pkg/k8s/types"
+	metallbspr "go.universe.tf/metallb/pkg/speaker"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const (
+	DefaultTimeout = 30 * time.Second
+)
+
+var (
+	errTimeout = errors.New("timeout occurred before mock received event")
+)
+
+// TestSpeakerOnUpdateService confirms the speaker performs the correct
+// actions when an OnUpdateService event takes place.
+func TestSpeakerOnUpdateService(t *testing.T) {
```

Comment:
Comment to hold on discussion at https://cilium.slack.com/archives/C2B917YHE/p1626113446459700